include/psa/crypto.h

```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
- *                              or required by the algorithm. This must be empty
- *                              if additional session data is not used by the
- *                              algorithm.
- * \param session_data_len      Size of the \p session_data buffer in bytes.
  *
  * \retval #PSA_SUCCESS
  *         Success.
  * \retval #PSA_ERROR_BAD_STATE
- *         The operation state is not valid (it must be inactive).
+ *         The operation state is not valid (it must have been set up.)
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
  * \retval #PSA_ERROR_INVALID_HANDLE
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_STORAGE_FAILURE
  * \retval #PSA_ERROR_NOT_PERMITTED
  * \retval #PSA_ERROR_INVALID_ARGUMENT
  *         \p key is not compatible with the algorithm in \p cipher_suite.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_pake_set_password_key(psa_pake_operation_t *operation,
+                                       mbedtls_svc_key_id_t password);
+
+/** Set the password for a password-authenticated key exchange via a memory hard
+ * function.
+ *
+ * Some protocols require using values derived from passwords via memory hard
+ * functions to mitigate dictionary attacks. Memory hard functions can be
+ * accessed through the key derivation interface and the result can be supplied
+ * to the PAKE operation in the form of a key derivation object.
+ *
+ * This function draws bytes from a key derivation algorithm and sets those
+ * bytes as a password for the password-authenticated key exchange.  If you
+ * view the key derivation's output as a stream of bytes, this function
+ * destructively reads the requested number of bytes from the stream.
+ * The key derivation operation's capacity decreases by the number of bytes read.
+ *
+ * If this function returns #PSA_ERROR_INVALID_ARGUMENT, \p key_derivation
+ * enters an error state and must be aborted by calling
+ * psa_key_derivation_abort().
+ *
+ * \param[in,out] operation       The operation object to set the password for.
+ *                                It must have been set up by psa_pake_setup()
+ *                                and not yet in use.
+ * \param[in,out] key_derivation  An ongoing key derivation operation set up
+ *                                from the password and in a state suitable for
+ *                                calling psa_key_derivation_output_bytes().
+ * \param input_length            Number of bytes to input from the
+ *                                \p key_derivation operation.
+ *
+ * \retval #PSA_SUCCESS
+ *         Success.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The operation state is not valid (it must have been set up.)
```

Comment:
> This description is for the PAKE operation, but there's also a KDF operation which must be in the right state as well.

In this case the call to `psa_key_derivation_output_bytes()` returns `PSA_ERROR_BAD_STATE` (which is something other than `PSA_ERROR_INSUFFICIENT_DATA`) and thus we return `PSA_ERROR_INVALID_ARGUMENT`. From the perspective of this function the KDF operation is an argument and the PAKE operation is the ongoing operation.

> For the PAKE, it's also an invalid state if psa_pake_output or psa_pake_input has already been called.

Good point, will fix.

Reply:
You have already stated that this function interacts with (and even modifies) two operation objects, and either could be in the wrong state. Cloaking a `BAD_STATE` error from the kdf does not help the application developer - as this is a key indicator that they have a bug in their code, rather than some transient issue.

Just about any error returned by the function leaves the two operation objects in an uncertain state, and the best advice will be for the application to abort both of them - it is hard to work out how any other approach would be reliable in this circumstance.

I expect that all of this commentary also applies to `psa_pake_get_implicit_key()`.

Reply:
> Just about any error returned by the function leaves the two operation objects in an uncertain state, and the best advice will be for the application to abort both of them - it is hard to work out how any other approach would be reliable in this circumstance.

Sounds good, will fix.